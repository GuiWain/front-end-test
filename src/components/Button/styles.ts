import styled, { css, DefaultTheme } from 'styled-components'

import { ButtonProps } from '.'

export type WrapperProps = {
  hasIcon?: boolean
} & Pick<ButtonProps, 'size' | 'fullWidth' | 'bg' | 'border' | 'iconPosition'>

export const Wrapper = styled.button<WrapperProps>`
  ${({
    theme,
    size,
    fullWidth,
    hasIcon,
    iconPosition,
    bg,
    border,
    disabled
  }) => css`
    display: inline-flex;
    align-items: center;
    gap: 0.6rem;
    justify-content: center;
    flex-direction: ${iconPosition === 'left' ? 'row' : 'row-reverse'};

    text-decoration: none;
    cursor: pointer;

    border-radius: ${theme.border.smallRadius};
    transition: all ${theme.transition.default};

    color: ${theme.colors.white};
    font-family: ${theme.font.family};
    font-weight: ${theme.font.bold};
    font-size: ${theme.font.sizes.xxsmall};
    line-height: 2.6rem;

    ${!!fullWidth && wrapperModifiers.fullWidth()};
    ${!!hasIcon && wrapperModifiers.withIcon()};
    ${disabled && wrapperModifiers.disabled(theme)};
    ${!!size && wrapperSizeModifiers[size](theme)};
    ${!!bg && wrapperBgModifiers[bg](theme)};
    ${!!border && wrapperBorderModifiers[border](theme)};
  `}
`

const wrapperModifiers = {
  fullWidth: () => css`
    width: 100%;
  `,
  withIcon: () => css`
    svg {
      width: 2rem;

      & + span {
        margin-left: 0.8rem;
      }
    }
  `,
  disabled: (theme: DefaultTheme) => css`
    &:disabled {
      color: ${theme.colors.neutralDark};
      cursor: not-allowed;
      filter: saturate(30%);
      background: ${theme.colors.neutralMedium} 0% 0% no-repeat padding-box;
      box-shadow: 0rem 0.4rem 0.8rem #00000014;
      border: none;

      :hover {
        background: ${theme.colors.neutralMedium} 0% 0% no-repeat padding-box;
        border: none;
        color: ${theme.colors.neutralDark};
      }
    }
  `
}

const wrapperSizeModifiers = {
  small: (theme: DefaultTheme) => css`
    height: 2.8rem;
    padding: 0.6rem 4.4rem;

    font-size: ${theme.font.sizes.xxxsmall};
    letter-spacing: 0rem;
  `,
  medium: (theme: DefaultTheme) => css`
    height: 3.6rem;
    padding: 0.9rem 4.4rem;

    font-size: ${theme.font.sizes.xxsmall};
    letter-spacing: 0rem;
  `,
  large: (theme: DefaultTheme) => css`
    height: 4.4rem;
    padding: 1.1rem 5.6rem;

    font-size: ${theme.font.sizes.xsmall};
    letter-spacing: 0rem;
  `,
  huge: (theme: DefaultTheme) => css`
    height: 5.2rem;
    padding: 1.4rem 5.6rem;

    font-size: ${theme.font.sizes.small};
    font-weight: ${theme.font.extraBold};
    line-height: ${theme.font.sizes.xlarge};
    letter-spacing: 0rem;
  `
}

const wrapperBgModifiers = {
  normal: (theme: DefaultTheme) => css`
    background: ${theme.colors.primaryMedium} 0% 0% no-repeat padding-box;
    box-shadow: 0rem 0.4rem 0.8rem #00000014;
    border: none;

    &:hover {
      background: ${theme.colors.primaryLight} 0% 0% no-repeat padding-box;
      transition: background ${theme.transition.default};
    }
  `,
  transparent: (theme: DefaultTheme) => css`
    color: ${theme.colors.secondaryDarkest};
    background: transparent 0% 0% no-repeat padding-box;
    border: 0.2rem solid ${theme.colors.primaryMedium};
    box-shadow: 0rem 0.4rem 0.8rem #00000014;

    &:hover {
      color: ${theme.colors.white};
      border-color: ${theme.colors.primaryLight};
      background: ${theme.colors.primaryLight} 0% 0% no-repeat padding-box;
      transition: background ${theme.transition.default};
    }
  `,
  gray: (theme: DefaultTheme) => css`
    color: ${theme.colors.secondaryDarkest};
    background: ${theme.colors.white} 0% 0% no-repeat padding-box;
    border: 0.1rem solid ${theme.colors.secondaryLight};

    &:hover {
      background: ${theme.colors.neutralLight} 0% 0% no-repeat padding-box;
      box-shadow: 0rem 0.4rem 0.8rem #00000014;
      transition: all ${theme.transition.default};
    }
  `,
  info: (theme: DefaultTheme) => css`
    color: ${theme.colors.white};
    background: ${theme.colors.infoDark} 0% 0% no-repeat padding-box;
    box-shadow: 0rem 0.4rem 0.8rem #00000014;
    border: none;

    &:hover {
      background: ${theme.colors.infoMedium} 0% 0% no-repeat padding-box;
      transition: all ${theme.transition.default};
    }
  `,
  success: (theme: DefaultTheme) => css`
    color: ${theme.colors.white};
    background: ${theme.colors.successMedium} 0% 0% no-repeat padding-box;
    box-shadow: 0rem 0.4rem 0.8rem #00000014;
    border: none;

    &:hover {
      background: ${theme.colors.successLight} 0% 0% no-repeat padding-box;
      transition: all ${theme.transition.default};
    }
  `,
  error: (theme: DefaultTheme) => css`
    color: ${theme.colors.white};
    background: ${theme.colors.errorMedium} 0% 0% no-repeat padding-box;
    box-shadow: 0rem 0.4rem 0.8rem #00000014;
    border: none;

    &:hover {
      background: ${theme.colors.errorLight} 0% 0% no-repeat padding-box;
      transition: all ${theme.transition.default};
    }
  `,
  warning: (theme: DefaultTheme) => css`
    color: ${theme.colors.white};
    background: ${theme.colors.warningDark} 0% 0% no-repeat padding-box;
    box-shadow: 0rem 0.4rem 0.8rem #00000014;
    border: none;

    &:hover {
      color: ${theme.colors.secondaryDarkest};
      background: ${theme.colors.warningMedium} 0% 0% no-repeat padding-box;
      transition: all ${theme.transition.default};
    }
  `
}

const wrapperBorderModifiers = {
  info: (theme: DefaultTheme) => css`
    color: ${theme.colors.infoDark};
    background: ${theme.colors.white} 0% 0% no-repeat padding-box;
    box-shadow: 0rem 0.4rem 0.8rem #00000014;
    border: 0.2rem solid ${theme.colors.infoDark};

    &:hover {
      color: ${theme.colors.white};
      border-color: ${theme.colors.infoMedium};
      background: ${theme.colors.infoMedium} 0% 0% no-repeat padding-box;
      transition: all ${theme.transition.default};
    }
  `,
  success: (theme: DefaultTheme) => css`
    color: ${theme.colors.successMedium};
    background: ${theme.colors.white} 0% 0% no-repeat padding-box;
    box-shadow: 0rem 0.4rem 0.8rem #00000014;
    border: 0.2rem solid ${theme.colors.successMedium};

    &:hover {
      color: ${theme.colors.white};
      border-color: ${theme.colors.successLight};
      background: ${theme.colors.successLight} 0% 0% no-repeat padding-box;
      transition: all ${theme.transition.default};
    }
  `,
  error: (theme: DefaultTheme) => css`
    color: ${theme.colors.errorMedium};
    background: ${theme.colors.white} 0% 0% no-repeat padding-box;
    box-shadow: 0rem 0.4rem 0.8rem #00000014;
    border: 0.2rem solid ${theme.colors.errorMedium};

    &:hover {
      color: ${theme.colors.white};
      border-color: ${theme.colors.errorLight};
      background: ${theme.colors.errorLight} 0% 0% no-repeat padding-box;
      transition: all ${theme.transition.default};
    }
  `,
  warning: (theme: DefaultTheme) => css`
    color: ${theme.colors.warningDark};
    background: ${theme.colors.white} 0% 0% no-repeat padding-box;
    box-shadow: 0rem 0.4rem 0.8rem #00000014;
    border: 0.2rem solid ${theme.colors.warningDark};

    &:hover {
      color: ${theme.colors.secondaryDarkest};
      border-color: ${theme.colors.warningMedium};
      background: ${theme.colors.warningMedium} 0% 0% no-repeat padding-box;
      transition: all ${theme.transition.default};
    }
  `
}
