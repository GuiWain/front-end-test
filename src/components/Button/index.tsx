import { forwardRef, AnchorHTMLAttributes, ButtonHTMLAttributes } from 'react'
import * as S from './styles'

type ButtonTypes =
  | AnchorHTMLAttributes<HTMLAnchorElement>
  | ButtonHTMLAttributes<HTMLButtonElement>

export type ButtonProps = {
  size?: 'small' | 'medium' | 'large' | 'huge'
  bg?:
    | 'normal'
    | 'transparent'
    | 'gray'
    | 'info'
    | 'success'
    | 'error'
    | 'warning'
  border?: 'info' | 'success' | 'error' | 'warning'
  fullWidth?: boolean
  icon?: JSX.Element
  as?: React.ElementType
} & ButtonTypes

const Button: React.ForwardRefRenderFunction<S.WrapperProps, ButtonProps> = (
  {
    children,
    icon,
    size = 'medium',
    fullWidth = false,
    bg = 'normal',
    border,
    ...props
  },
  ref
) => (
  <S.Wrapper
    size={size}
    fullWidth={fullWidth}
    hasIcon={!!icon}
    bg={bg}
    border={border}
    ref={ref}
    {...props}
  >
    {icon}
    {!!children && children}
  </S.Wrapper>
)

export default forwardRef(Button)
