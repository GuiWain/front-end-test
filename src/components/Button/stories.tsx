import { Story, Meta } from '@storybook/react/types-6-0'
import { AddShoppingCart } from '@styled-icons/material-outlined/AddShoppingCart'
import Button, { ButtonProps } from '.'

export default {
  title: 'Components/Button',
  component: Button,
  argTypes: {
    children: {
      type: 'string'
    },
    icon: {
      type: ''
    },
    size: {
      control: {
        type: 'radio',
        options: ['small', 'medium', 'large', 'huge']
      }
    },
    disabled: {
      control: {
        type: 'radio',
        options: [true, false]
      }
    },
    fullWidth: {
      control: {
        type: 'radio',
        options: [true, false]
      }
    },
    bg: {
      control: {
        type: 'radio',
        options: [
          'normal',
          'transparent',
          'gray',
          'info',
          'success',
          'error',
          'warning'
        ]
      }
    },
    border: {
      control: {
        type: 'radio',
        options: ['info', 'success', 'error', 'warning']
      }
    }
  }
} as Meta

export const Default: Story<ButtonProps> = (args) => <Button {...args} />

Default.args = {
  size: 'medium',
  children: 'Click',
  bg: 'normal'
}

export const WithBorder: Story<ButtonProps> = (args) => <Button {...args} />

WithBorder.args = {
  size: 'medium',
  children: 'Click',
  border: 'info'
}

export const withIcon: Story<ButtonProps> = (args) => <Button {...args} />

withIcon.args = {
  size: 'medium',
  children: 'Click',
  icon: <AddShoppingCart />
}
