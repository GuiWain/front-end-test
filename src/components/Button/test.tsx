import { render, screen } from 'utils/test-utils'
import { AddShoppingCart } from '@styled-icons/material-outlined/AddShoppingCart'

import Button from '.'

describe('<Button />', () => {
  it('should render Button by default', () => {
    const { container } = render(<Button>Click</Button>)

    const button = screen.getByRole('button', { name: /Click/i })
    expect(button).toHaveStyle({
      height: '3.6rem',
      padding: '0.9rem 4.4rem',
      background: '#E1E974 0% 0% no-repeat padding-box',
      'font-size': '1.4rem'
    })

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render the small size', () => {
    render(<Button size="small">Click</Button>)

    const button = screen.getByRole('button', { name: /Click/i })
    expect(button).toHaveStyle({
      height: '2.8rem',
      padding: '0.6rem 4.4rem',
      'font-size': '1.2rem'
    })
  })

  it('should render the large size', () => {
    render(<Button size="large">Click</Button>)

    const button = screen.getByRole('button', { name: /Click/i })
    expect(button).toHaveStyle({
      height: '4.4rem',
      padding: '1.1rem 5.6rem',
      'font-size': '1.6rem'
    })
  })

  it('should render the huge size', () => {
    render(<Button size="huge">Click</Button>)

    const button = screen.getByRole('button', { name: /Click/i })
    expect(button).toHaveStyle({
      height: '5.2rem',
      padding: '1.4rem 5.6rem',
      'font-size': '1.8rem',
      'font-weight': '800'
    })
  })

  it('should render a bg transparent version', () => {
    render(<Button bg="transparent">Click</Button>)

    const button = screen.getByRole('button', { name: /Click/i })
    expect(button).toHaveStyle({
      background: 'transparent 0% 0% no-repeat padding-box',
      border: '0.2rem solid #510161'
    })
  })

  it('should render a bg gray version', () => {
    render(<Button bg="gray">Click</Button>)

    const button = screen.getByRole('button', { name: /Click/i })
    expect(button).toHaveStyle({
      background: '#FFFFFF 0% 0% no-repeat padding-box',
      border: '0.1rem solid #CCC8C8'
    })
  })

  it('should render a bg info version', () => {
    render(<Button bg="info">Click</Button>)

    const button = screen.getByRole('button', { name: /Click/i })
    expect(button).toHaveStyle({
      background: '#3241A1 0% 0% no-repeat padding-box',
      border: 'none',
      color: '#FFFFFF'
    })
  })

  it('should render a bg success version', () => {
    render(<Button bg="success">Click</Button>)

    const button = screen.getByRole('button', { name: /Click/i })
    expect(button).toHaveStyle({
      background: '#00A03E 0% 0% no-repeat padding-box',
      border: 'none',
      color: '#FFFFFF'
    })
  })

  it('should render a bg error version', () => {
    render(<Button bg="error">Click</Button>)

    const button = screen.getByRole('button', { name: /Click/i })
    expect(button).toHaveStyle({
      background: '#ed3020 0% 0% no-repeat padding-box',
      border: 'none',
      color: '#FFFFFF'
    })
  })

  it('should render a bg warning version', () => {
    render(<Button bg="warning">Click</Button>)

    const button = screen.getByRole('button', { name: /Click/i })
    expect(button).toHaveStyle({
      background: '#e7b500 0% 0% no-repeat padding-box',
      border: 'none',
      color: '#FFFFFF'
    })
  })

  it('should render a border info version', () => {
    render(<Button border="info">Click</Button>)

    const button = screen.getByRole('button', { name: /Click/i })
    expect(button).toHaveStyle({
      background: '#FFFFFF 0% 0% no-repeat padding-box',
      border: '0.2rem solid #3241A1',
      color: '#3241A1'
    })
  })

  it('should render a border success version', () => {
    render(<Button border="success">Click</Button>)

    const button = screen.getByRole('button', { name: /Click/i })
    expect(button).toHaveStyle({
      background: '#FFFFFF 0% 0% no-repeat padding-box',
      border: '0.2rem solid #00A03E',
      color: '#00A03E'
    })
  })

  it('should render a border error version', () => {
    render(<Button border="error">Click</Button>)

    const button = screen.getByRole('button', { name: /Click/i })
    expect(button).toHaveStyle({
      background: '#FFFFFF 0% 0% no-repeat padding-box',
      border: '0.2rem solid #ed3020',
      color: '#ed3020'
    })
  })

  it('should render a border warning version', () => {
    render(<Button border="warning">Click</Button>)

    const button = screen.getByRole('button', { name: /Click/i })
    expect(button).toHaveStyle({
      background: '#FFFFFF 0% 0% no-repeat padding-box',
      border: '0.2rem solid #e7b500',
      color: '#e7b500'
    })
  })

  it('should render a fullWidth version', () => {
    render(<Button fullWidth>Click</Button>)

    expect(screen.getByRole('button', { name: /Click/i })).toHaveStyle({
      width: '100%'
    })
  })

  it('should render an icon version', () => {
    render(<Button icon={<AddShoppingCart data-testid="icon" />}>Click</Button>)

    expect(screen.getByText(/Click/i)).toBeInTheDocument()
    expect(screen.getByTestId('icon')).toBeInTheDocument()
  })

  it('should render a disabled Button', () => {
    render(<Button disabled>Click</Button>)

    const button = screen.getByRole('button', { name: /Click/i })
    expect(button).toHaveStyleRule('cursor', 'not-allowed', {
      modifier: ':disabled'
    })

    expect(button).toHaveStyle({
      background: '#D9D9D9 0% 0% no-repeat padding-box',
      border: 'none',
      color: '#9D9D9D'
    })
  })
})
