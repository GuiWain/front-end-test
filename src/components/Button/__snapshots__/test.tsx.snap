// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render Button by default 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 0.6rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  border-radius: 0.4rem;
  -webkit-transition: all 0.5s ease-in-out;
  transition: all 0.5s ease-in-out;
  color: #FFFFFF;
  font-family: 'Open Sans','Roboto','Helvetica Neue',sans-serif;
  font-weight: 700;
  font-size: 1.4rem;
  line-height: 2.6rem;
  height: 3.6rem;
  padding: 0.9rem 4.4rem;
  font-size: 1.4rem;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  background: #510161 0% 0% no-repeat padding-box;
  box-shadow: 0rem 0.4rem 0.8rem #00000014;
  border: none;
}

.c0:hover {
  background: #8802A3 0% 0% no-repeat padding-box;
  -webkit-transition: background 0.5s ease-in-out;
  transition: background 0.5s ease-in-out;
}

<button
  class="c0"
>
  Click
</button>
`;
